src/graphics/primitives/collision/shader.rs

```diff
@@ -0,0 +1,155 @@
+// Copyright (C) 2023 Lily Lyons
+//
+// This file is part of Luminol.
+//
+// Luminol is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Luminol is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Luminol.  If not, see <http://www.gnu.org/licenses/>.
+use crate::prelude::*;
+
+use super::instance::Instances;
+use super::Vertex;
+use const_format::str_replace;
+use primitives::Viewport;
+
+#[derive(Debug)]
+pub struct Shader {
+    pub pipeline: wgpu::RenderPipeline,
+}
+
+impl Shader {
+    fn new(use_push_constants: bool) -> Self {
+        let render_state = &state!().render_state;
+
+        let shader_module = if use_push_constants {
+            render_state
+                .device
+                .create_shader_module(wgpu::ShaderModuleDescriptor {
+                    label: Some("collision.wgsl (push constants)"),
+                    source: wgpu::ShaderSource::Wgsl(
+                        str_replace!(
+                            concat!(
+                                include_str!("collision_header_push_constants.wgsl"),
+                                include_str!("collision.wgsl"),
+                            ),
+                            "HOST.",
+                            "push_constants."
+                        )
+                        .into(),
+                    ),
+                })
+        } else {
+            render_state
+                .device
+                .create_shader_module(wgpu::ShaderModuleDescriptor {
+                    label: Some("collision.wgsl (uniforms)"),
+                    source: wgpu::ShaderSource::Wgsl(
+                        str_replace!(
+                            concat!(
+                                include_str!("collision_header_uniforms.wgsl"),
+                                include_str!("collision.wgsl"),
+                            ),
+                            "HOST.",
+                            ""
+                        )
+                        .into(),
+                    ),
+                })
+        };
+
+        let pipeline_layout = if use_push_constants {
+            render_state
+                .device
+                .create_pipeline_layout(&wgpu::PipelineLayoutDescriptor {
+                    label: Some("Tilemap Collision Render Pipeline Layout (push constants)"),
+                    bind_group_layouts: &[&DUMMY_LAYOUT],
```

Comment:
Why is there a dummy layout?

Reply:
I was using that for testing purposes because the viewport is hardcoded to bind to group 1 and then didn't remove it. I've just changed the viewport to allow specifying a group index.